$\large{\textbf{2.4 Formulation:}} \\ \text{max } 11x_1 + 13x_2 + 9x_3 + 15x_4 + 12x_5 \\ \text{s.t. } 6.5x_1 + 8x_2 + 10.5x_3 + 11.75x_4 + 14.5x_5 \leq 78 \\ 5.75x_1 + 4.5x_2 + 3.75x_3 + 4.85x_4 + 3.45x_5 \leq 84 \\ x_i \in 𝐙 \ ∀ \ i = \{1,2,3,4,5\} \\ \text{where } x_i \text{'s are the number of drones bought of type i}$

In [245]:
!pip install -q pyomo

In [246]:
from pyomo.environ import *

In [247]:
import numpy as np

In [248]:
import pandas as pd

In [249]:
data_csvfile = pd.read_csv('lab4_ex3.csv')

In [250]:
print(type(data_csvfile))

<class 'pandas.core.frame.DataFrame'>


In [251]:
data_csvfile.columns

Index(['Drone Model', 'Capacity', 'Cost', 'Parking Space'], dtype='object')

In [252]:
data_csvfile.index

RangeIndex(start=0, stop=5, step=1)

In [253]:
data_csvfile[1:2]

Drone Model  Capacity  Cost  Parking Space
1      Dharan        13   8.0            4.5

In [254]:
data_csvfile['Drone Model']

0     Shrey
1    Dharan
2     Satva
3     Adhya
4    Pravat
Name: Drone Model, dtype: object

In [255]:
N = len(data_csvfile.index)

In [256]:
M = 2 #no of constraints

In [257]:
col_indices = np.arange(N)

In [258]:
model_ex2 = ConcreteModel()

In [259]:
model_ex2.x = Var(col_indices,domain=NonNegativeIntegers)

$\text{Here all the variables are taken integers as the number of drones bought cannot be in floating type.}$

In [260]:
model_ex2.obj = Objective(expr = summation(data_csvfile['Capacity'],model_ex2.x),sense=maximize)

In [261]:
model_ex2.const = ConstraintList()

In [262]:
model_ex2.const.add(expr = summation(data_csvfile['Cost'],model_ex2.x) <= 78)
model_ex2.const.add(expr = summation(data_csvfile['Parking Space'],model_ex2.x) <= 84)

In [263]:
model_ex2.pprint()

2 Set Declarations
    const_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    2 : {1, 2}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  None :  None : False :  True : NonNegativeIntegers
          1 :     0 :  None :  None : False :  True : NonNegativeIntegers
          2 :     0 :  None :  None : False :  True : NonNegativeIntegers
          3 :     0 :  None :  None : False :  True : NonNegativeIntegers
          4 :     0 :  None :  None : False :  True : NonNegativeIntegers

1 Objective Declarations
    obj : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 11*x[0] + 13*x[1] + 9*x[2] + 15*x[3] + 12*x[4]

1 Co

In [264]:
!apt-get install -y -qq coinor-cbc

In [265]:
opt_cbc = SolverFactory('cbc')

In [266]:
result = opt_cbc.solve(model_ex2)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [267]:
print('\nObjective = ', model_ex2.obj())

print('\nDecision Variables')
for i in col_indices:
  print('Number of Drone',data_csvfile['Drone Model'][i], model_ex2.x[i].value)

print('\nConstraints')
model_ex2.const.display()


Objective =  132.0

Decision Variables
Number of Drone Shrey 12.0
Number of Drone Dharan 0.0
Number of Drone Satva 0.0
Number of Drone Adhya 0.0
Number of Drone Pravat 0.0

Constraints
const : Size=2
    Key : Lower : Body : Upper
      1 :  None : 78.0 :  78.0
      2 :  None : 69.0 :  84.0


$\large{\textbf{Answer 2.6 :}} \\ \text{Maximum carrying capacity/Objective Function value = } 132 \text{ kgs.} \\ \text{Value of decision variables:} \\ x_1 = 12 \\ x_2 =0 \\ x_3 = 0 \\ x_4 =0 \\ x_5 = 0 $

In [268]:
model_ex2.x.domain=NonNegativeReals

In [269]:
model_ex2.pprint()

2 Set Declarations
    const_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    2 : {1, 2}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  12.0 :  None : False : False : NonNegativeReals
          1 :     0 :   0.0 :  None : False : False : NonNegativeReals
          2 :     0 :   0.0 :  None : False : False : NonNegativeReals
          3 :     0 :   0.0 :  None : False : False : NonNegativeReals
          4 :     0 :   0.0 :  None : False : False : NonNegativeReals

1 Objective Declarations
    obj : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 11*x[0] + 13*x[1] + 9*x[2] + 15*x[3] + 12*x[4]

1 Constraint Declar

In [270]:
result = opt_cbc.solve(model_ex2)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [271]:
print('\nObjective = ', model_ex2.obj())

print('\nDecision Variables')
for i in col_indices:
  print('Number of Drone',data_csvfile['Drone Model'][i], model_ex2.x[i].value)

print('\nConstraints')
model_ex2.const.display()


Objective =  132.0

Decision Variables
Number of Drone Shrey 12.0
Number of Drone Dharan 0.0
Number of Drone Satva 0.0
Number of Drone Adhya 0.0
Number of Drone Pravat 0.0

Constraints
const : Size=2
    Key : Lower : Body : Upper
      1 :  None : 78.0 :  78.0
      2 :  None : 69.0 :  84.0


$\large{\textbf{Answer 2.7 :}} \\ \text{Maximum carrying capacity/Objective Function value = } 132 \text{ kgs.} \\ \text{Value of decision variables:} \\ x_1 = 12 \\ x_2 =0 \\ x_3 = 0 \\ x_4 =0 \\ x_5 = 0 $

$\large{\textbf{Answer 2.8 : }} \\ \text{No,in general the solution of the MILP cannot be obtained by merely rounding off the solution of the Linear Program.} \\ \text{There can be cases where the solution obtained by the process are feasible but far off from the optimal solution} \\ \text{and in some cases after rounding off the solution might also get out of the feasible region.} \\ \text{In this case the solution is coming out to be same as the cost constraint is active} \\ \text{i.e. the cost is completely utilized so no new drones can be bought in order to increase the carrying capacity.}$

$\large{\textbf{Answer 2.9 :}} \\ \text{Here we are reducing the budget to 76 crores and parking space is increased to 87 sq. units.}$

In [272]:
model_ex2.const[1].deactivate()
model_ex2.const[2].deactivate()
model_ex2.const.add(expr = summation(data_csvfile['Cost'],model_ex2.x) <= 76)
model_ex2.const.add(expr = summation(data_csvfile['Parking Space'],model_ex2.x) <= 87)

In [273]:
model_ex2.pprint()

2 Set Declarations
    const_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    4 : {1, 2, 3, 4}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  12.0 :  None : False : False : NonNegativeReals
          1 :     0 :   0.0 :  None : False : False : NonNegativeReals
          2 :     0 :   0.0 :  None : False : False : NonNegativeReals
          3 :     0 :   0.0 :  None : False : False : NonNegativeReals
          4 :     0 :   0.0 :  None : False : False : NonNegativeReals

1 Objective Declarations
    obj : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 11*x[0] + 13*x[1] + 9*x[2] + 15*x[3] + 12*x[4]

1 Constraint 

In [274]:
result = opt_cbc.solve(model_ex2)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [275]:
print('\nObjective = ', model_ex2.obj())

print('\nDecision Variables')
for i in col_indices:
  print('Number of Drone',data_csvfile['Drone Model'][i], model_ex2.x[i].value)

print('\nConstraints')
model_ex2.const.display()


Objective =  128.615388

Decision Variables
Number of Drone Shrey 11.692308
Number of Drone Dharan 0.0
Number of Drone Satva 0.0
Number of Drone Adhya 0.0
Number of Drone Pravat 0.0

Constraints
const : Size=4
    Key : Lower : Body              : Upper
      3 :  None : 76.00000200000001 :  76.0
      4 :  None :         67.230771 :  87.0


$\large{\textbf{For LP:}} \\ \text{Maximum carrying capacity/Objective Function value = } 128.615388 \text{ kgs.} \\ \text{Value of decision variables:} \\ x_1 = 11.692308 \\ x_2 =0 \\ x_3 = 0 \\ x_4 =0 \\ x_5 = 0 $

In [276]:
model_ex2.x.domain=NonNegativeIntegers

In [277]:
result = opt_cbc.solve(model_ex2)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [278]:
print('\nObjective = ', model_ex2.obj())

print('\nDecision Variables')
for i in col_indices:
  print('Number of Drone',data_csvfile['Drone Model'][i], model_ex2.x[i].value)

print('\nConstraints')
model_ex2.const.display()


Objective =  127.0

Decision Variables
Number of Drone Shrey 8.0
Number of Drone Dharan 3.0
Number of Drone Satva 0.0
Number of Drone Adhya 0.0
Number of Drone Pravat 0.0

Constraints
const : Size=4
    Key : Lower : Body : Upper
      3 :  None : 76.0 :  76.0
      4 :  None : 59.5 :  87.0


$\large{\textbf{For MILP:}} \\ \text{Maximum carrying capacity/Objective Function value = } 127 \text{ kgs.} \\ \text{Value of decision variables:} \\ x_1 = 8 \\ x_2 =3 \\ x_3 = 0 \\ x_4 =0 \\ x_5 = 0 \\ \text{Here we can observe that we get different solutions on rounding off the solution of the LP and the modified MILP.}\\ \text{Hence in general rounding off the solutions of the LP to get the solution of the MILP is incorrect.}$

$\text{Here we can observe that the budget constraint is acting as the dominant constraint.} \\ \text{On reducing the budget, we observe that the maximum carrying capacity is also decreasing in both the cases.}$

$\large{\textbf{Answer 2.10 : }} \\ \textbf{Case 1: } \\ \text{Budget Constraint = 74 crores } \\ \text{Parking Space constraint = 88 sq. m.}$

In [279]:
model_ex2.const[3].deactivate()
model_ex2.const[4].deactivate()
model_ex2.const.add(expr = summation(data_csvfile['Cost'],model_ex2.x) <= 74)
model_ex2.const.add(expr = summation(data_csvfile['Parking Space'],model_ex2.x) <= 88)

In [280]:
model_ex2.pprint()

2 Set Declarations
    const_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    6 : {1, 2, 3, 4, 5, 6}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :   8.0 :  None : False : False : NonNegativeIntegers
          1 :     0 :   3.0 :  None : False : False : NonNegativeIntegers
          2 :     0 :   0.0 :  None : False : False : NonNegativeIntegers
          3 :     0 :   0.0 :  None : False : False : NonNegativeIntegers
          4 :     0 :   0.0 :  None : False : False : NonNegativeIntegers

1 Objective Declarations
    obj : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 11*x[0] + 13*x[1] + 9*x[2] + 15*x[3] + 1

In [281]:
result = opt_cbc.solve(model_ex2)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [282]:
print('\nObjective = ', model_ex2.obj())

print('\nDecision Variables')
for i in col_indices:
  print('Number of Drone',data_csvfile['Drone Model'][i], model_ex2.x[i].value)

print('\nConstraints')
model_ex2.const.display()


Objective =  123.0

Decision Variables
Number of Drone Shrey 10.0
Number of Drone Dharan 1.0
Number of Drone Satva 0.0
Number of Drone Adhya 0.0
Number of Drone Pravat 0.0

Constraints
const : Size=6
    Key : Lower : Body : Upper
      5 :  None : 73.0 :  74.0
      6 :  None : 62.0 :  88.0


$\large{\textbf{For MILP:}} \\ \text{Maximum carrying capacity/Objective Function value = } 123 \text{ kgs.} \\ \text{Value of decision variables:} \\ x_1 = 10 \\ x_2 =1 \\ x_3 = 0 \\ x_4 =0 \\ x_5 = 0 $

In [283]:
model_ex2.x.domain=NonNegativeReals

In [284]:
result = opt_cbc.solve(model_ex2)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [285]:
print('\nObjective = ', model_ex2.obj())

print('\nDecision Variables')
for i in col_indices:
  print('Number of Drone',data_csvfile['Drone Model'][i], model_ex2.x[i].value)

print('\nConstraints')
model_ex2.const.display()


Objective =  125.230765

Decision Variables
Number of Drone Shrey 11.384615
Number of Drone Dharan 0.0
Number of Drone Satva 0.0
Number of Drone Adhya 0.0
Number of Drone Pravat 0.0

Constraints
const : Size=6
    Key : Lower : Body        : Upper
      5 :  None :  73.9999975 :  74.0
      6 :  None : 65.46153625 :  88.0


$\large{\textbf{For LP:}} \\ \text{Maximum carrying capacity/Objective Function value = } 125.230765 \text{ kgs.} \\ \text{Value of decision variables:} \\ x_1 = 11.384615 \\ x_2 =0 \\ x_3 = 0 \\ x_4 =0 \\ x_5 = 0 $

$\textbf{Case 2: } \\ \text{Budget Constraint = 72 crores } \\ \text{Parking Space constraint = 89 sq. m.}$

In [286]:
model_ex2.const[5].deactivate()
model_ex2.const[6].deactivate()
model_ex2.const.add(expr = summation(data_csvfile['Cost'],model_ex2.x) <= 72)
model_ex2.const.add(expr = summation(data_csvfile['Parking Space'],model_ex2.x) <= 89)

In [287]:
model_ex2.pprint()

2 Set Declarations
    const_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    8 : {1, 2, 3, 4, 5, 6, 7, 8}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value     : Upper : Fixed : Stale : Domain
          0 :     0 : 11.384615 :  None : False : False : NonNegativeReals
          1 :     0 :       0.0 :  None : False : False : NonNegativeReals
          2 :     0 :       0.0 :  None : False : False : NonNegativeReals
          3 :     0 :       0.0 :  None : False : False : NonNegativeReals
          4 :     0 :       0.0 :  None : False : False : NonNegativeReals

1 Objective Declarations
    obj : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 11*x[0] + 13*x[1] + 9*x[2

In [288]:
result = opt_cbc.solve(model_ex2)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [289]:
print('\nObjective = ', model_ex2.obj())

print('\nDecision Variables')
for i in col_indices:
  print('Number of Drone',data_csvfile['Drone Model'][i], model_ex2.x[i].value)

print('\nConstraints')
model_ex2.const.display()


Objective =  121.84615300000002

Decision Variables
Number of Drone Shrey 11.076923
Number of Drone Dharan 0.0
Number of Drone Satva 0.0
Number of Drone Adhya 0.0
Number of Drone Pravat 0.0

Constraints
const : Size=8
    Key : Lower : Body               : Upper
      7 :  None :         71.9999995 :  72.0
      8 :  None : 63.692307250000006 :  89.0


$\large{\textbf{For LP:}} \\ \text{Maximum carrying capacity/Objective Function value = } 121.8462 \text{ kgs.} \\ \text{Value of decision variables:} \\ x_1 = 11.0769 \\ x_2 =0 \\ x_3 = 0 \\ x_4 =0 \\ x_5 = 0 $

In [290]:
model_ex2.x.domain=NonNegativeIntegers

In [291]:
result = opt_cbc.solve(model_ex2)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [292]:
print('\nObjective = ', model_ex2.obj())

print('\nDecision Variables')
for i in col_indices:
  print('Number of Drone',data_csvfile['Drone Model'][i], model_ex2.x[i].value)

print('\nConstraints')
model_ex2.const.display()


Objective =  121.0

Decision Variables
Number of Drone Shrey 11.0
Number of Drone Dharan 0.0
Number of Drone Satva 0.0
Number of Drone Adhya 0.0
Number of Drone Pravat 0.0

Constraints
const : Size=8
    Key : Lower : Body  : Upper
      7 :  None :  71.5 :  72.0
      8 :  None : 63.25 :  89.0


$\large{\textbf{For MILP:}} \\ \text{Maximum carrying capacity/Objective Function value = } 121 \text{ kgs.} \\ \text{Value of decision variables:} \\ x_1 = 11 \\ x_2 =0 \\ x_3 = 0 \\ x_4 =0 \\ x_5 = 0 $

$\textbf{Case 3: } \\ \text{Budget Constraint = 70 crores } \\ \text{Parking Space constraint = 90 sq. m.}$

In [293]:
model_ex2.const[7].deactivate()
model_ex2.const[8].deactivate()
model_ex2.const.add(expr = summation(data_csvfile['Cost'],model_ex2.x) <= 70)
model_ex2.const.add(expr = summation(data_csvfile['Parking Space'],model_ex2.x) <= 90)

In [294]:
model_ex2.pprint()

2 Set Declarations
    const_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   10 : {1, 2, 3, 4, 5, 6, 7, 8, 9, 10}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  11.0 :  None : False : False : NonNegativeIntegers
          1 :     0 :   0.0 :  None : False : False : NonNegativeIntegers
          2 :     0 :   0.0 :  None : False : False : NonNegativeIntegers
          3 :     0 :   0.0 :  None : False : False : NonNegativeIntegers
          4 :     0 :   0.0 :  None : False : False : NonNegativeIntegers

1 Objective Declarations
    obj : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 11*x[0] + 13*x[1] + 9*x[2] 

In [295]:
result = opt_cbc.solve(model_ex2)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [296]:
print('\nObjective = ', model_ex2.obj())

print('\nDecision Variables')
for i in col_indices:
  print('Number of Drone',data_csvfile['Drone Model'][i], model_ex2.x[i].value)

print('\nConstraints')
model_ex2.const.display()


Objective =  116.0

Decision Variables
Number of Drone Shrey 7.0
Number of Drone Dharan 3.0
Number of Drone Satva 0.0
Number of Drone Adhya 0.0
Number of Drone Pravat 0.0

Constraints
const : Size=10
    Key : Lower : Body  : Upper
      9 :  None :  69.5 :  70.0
     10 :  None : 53.75 :  90.0


$\large{\textbf{For MILP:}} \\ \text{Maximum carrying capacity/Objective Function value = } 116 \text{ kgs.} \\ \text{Value of decision variables:} \\ x_1 = 7 \\ x_2 =3 \\ x_3 = 0 \\ x_4 =0 \\ x_5 = 0 $

In [297]:
model_ex2.x.domain = NonNegativeReals

In [298]:
result = opt_cbc.solve(model_ex2)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [299]:
print('\nObjective = ', model_ex2.obj())

print('\nDecision Variables')
for i in col_indices:
  print('Number of Drone',data_csvfile['Drone Model'][i], model_ex2.x[i].value)

print('\nConstraints')
model_ex2.const.display()


Objective =  118.461541

Decision Variables
Number of Drone Shrey 10.769231
Number of Drone Dharan 0.0
Number of Drone Satva 0.0
Number of Drone Adhya 0.0
Number of Drone Pravat 0.0

Constraints
const : Size=10
    Key : Lower : Body               : Upper
      9 :  None :         70.0000015 :  70.0
     10 :  None : 61.923078249999996 :  90.0


$\large{\textbf{For LP:}} \\ \text{Maximum carrying capacity/Objective Function value = } 118.4615 \text{ kgs.} \\ \text{Value of decision variables:} \\ x_1 = 10.769 \\ x_2 =3 \\ x_3 = 0 \\ x_4 =0 \\ x_5 = 0 $

$\textbf{Case 4: } \\ \text{Budget Constraint = 68 crores } \\ \text{Parking Space constraint = 91 sq. m.}$

In [300]:
model_ex2.const[9].deactivate()
model_ex2.const[10].deactivate()
model_ex2.const.add(expr = summation(data_csvfile['Cost'],model_ex2.x) <= 68)
model_ex2.const.add(expr = summation(data_csvfile['Parking Space'],model_ex2.x) <= 91)

In [301]:
model_ex2.pprint()

2 Set Declarations
    const_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   12 : {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value     : Upper : Fixed : Stale : Domain
          0 :     0 : 10.769231 :  None : False : False : NonNegativeReals
          1 :     0 :       0.0 :  None : False : False : NonNegativeReals
          2 :     0 :       0.0 :  None : False : False : NonNegativeReals
          3 :     0 :       0.0 :  None : False : False : NonNegativeReals
          4 :     0 :       0.0 :  None : False : False : NonNegativeReals

1 Objective Declarations
    obj : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 11*x[0] + 

In [302]:
result = opt_cbc.solve(model_ex2)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [303]:
print('\nObjective = ', model_ex2.obj())

print('\nDecision Variables')
for i in col_indices:
  print('Number of Drone',data_csvfile['Drone Model'][i], model_ex2.x[i].value)

print('\nConstraints')
model_ex2.const.display()


Objective =  115.07691799999999

Decision Variables
Number of Drone Shrey 10.461538
Number of Drone Dharan 0.0
Number of Drone Satva 0.0
Number of Drone Adhya 0.0
Number of Drone Pravat 0.0

Constraints
const : Size=12
    Key : Lower : Body               : Upper
     11 :  None :          67.999997 :  68.0
     12 :  None : 60.153843499999994 :  91.0


$\large{\textbf{For LP:}} \\ \text{Maximum carrying capacity/Objective Function value = } 115.0769 \text{ kgs.} \\ \text{Value of decision variables:} \\ x_1 = 10.461538 \\ x_2 = 0 \\ x_3 = 0 \\ x_4 =0 \\ x_5 = 0 $

In [304]:
model_ex2.x.domain=NonNegativeIntegers

In [305]:
result = opt_cbc.solve(model_ex2)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [306]:
print('\nObjective = ', model_ex2.obj())

print('\nDecision Variables')
for i in col_indices:
  print('Number of Drone',data_csvfile['Drone Model'][i], model_ex2.x[i].value)

print('\nConstraints')
model_ex2.const.display()


Objective =  114.0

Decision Variables
Number of Drone Shrey 8.0
Number of Drone Dharan 2.0
Number of Drone Satva 0.0
Number of Drone Adhya 0.0
Number of Drone Pravat 0.0

Constraints
const : Size=12
    Key : Lower : Body : Upper
     11 :  None : 68.0 :  68.0
     12 :  None : 55.0 :  91.0


$\large{\textbf{For MILP:}} \\ \text{Maximum carrying capacity/Objective Function value = } 114 \text{ kgs.} \\ \text{Value of decision variables:} \\ x_1 = 8 \\ x_2 =2 \\ x_3 = 0 \\ x_4 =0 \\ x_5 = 0 $

$\text{Here we can observe that the budget constraint is acting as the dominant constraint.} \\ \text{On reducing the budget, we observe that the maximum carrying capacity is also decreasing in both the cases.} \\ \text{In ideal case increment in the parking space constraint should have balanced the effect of the budget constraint.} \\ \text{But that is not the case, the objective function value is decreasing as the budget constraint decreases.} \\ \textbf{Observations on optimal solutions obtained:} \\ \text{When there is no constraint on the variables being integers,} \\ \text{the model is suggesting to buy as many} \textbf{ Shrey Drones} \text{ as possible as it is the cheapest and maybe serves our purpose the best.} \\ \text{But in case of MILP, we buy either } \textbf{Shrey drones,} \text{or } \textbf{Dharan drones.}$